File: go/0654M-最大二叉树.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-04 15:46:39
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructMaximumBinaryTree(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	rootIndex := 0
	for i, v := range nums {
		if v > nums[rootIndex] {
			rootIndex = i
		}
	}
	root := &TreeNode{
		Val: nums[rootIndex],
	}
	if len(nums) == 1 {
		return root
	}

	root.Left = constructMaximumBinaryTree(nums[:rootIndex])
	root.Right = constructMaximumBinaryTree(nums[rootIndex+1:])
	return root
}

File: cpp/0654M-最大二叉树.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-21 15:31:33
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        if (nums.empty()) return nullptr;
        int maxIndex = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > nums[maxIndex]) maxIndex = i;
        }
        TreeNode* root = new TreeNode(nums[maxIndex]);
        if (maxIndex > 0) {
            vector<int> left(nums.begin(), nums.begin() + maxIndex);
            root->left = constructMaximumBinaryTree(left);
        }
        if (maxIndex < nums.size() - 1) {
            vector<int> right(nums.begin() + maxIndex + 1, nums.end());
            root->right = constructMaximumBinaryTree(right);
        }
        return root;
    }
};
